// angular2-baidu-map/esm2015/types/Circle.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function BCircleConstructor() { }
function BCircleConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (center: BPoint, radius: number, options?: CircleOptions): BCircle
    */
}
/**
 * @record
 */
export function BCircle() { }
function BCircle_tsickle_Closure_declarations() {
    /** @type {?} */
    BCircle.prototype.setCenter;
    /** @type {?} */
    BCircle.prototype.setRadius;
    /** @type {?} */
    BCircle.prototype.setStrokeColor;
    /** @type {?} */
    BCircle.prototype.setFillColor;
    /** @type {?} */
    BCircle.prototype.setStrokeOpacity;
    /** @type {?} */
    BCircle.prototype.setFillOpacity;
    /** @type {?} */
    BCircle.prototype.setStrokeWeight;
    /** @type {?} */
    BCircle.prototype.setStrokeStyle;
    /** @type {?} */
    BCircle.prototype.enableEditing;
    /** @type {?} */
    BCircle.prototype.disableEditing;
    /** @type {?} */
    BCircle.prototype.enableMassClear;
    /** @type {?} */
    BCircle.prototype.disableMassClear;
}
/**
 * @record
 */
export function CircleOptions() { }
function CircleOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    CircleOptions.prototype.strokeColor;
    /** @type {?|undefined} */
    CircleOptions.prototype.fillColor;
    /** @type {?|undefined} */
    CircleOptions.prototype.strokeWeight;
    /** @type {?|undefined} */
    CircleOptions.prototype.strokeOpacity;
    /** @type {?|undefined} */
    CircleOptions.prototype.fillOpacity;
    /** @type {?|undefined} */
    CircleOptions.prototype.strokeStyle;
    /** @type {?|undefined} */
    CircleOptions.prototype.enableMassClear;
    /** @type {?|undefined} */
    CircleOptions.prototype.enableEditing;
    /** @type {?|undefined} */
    CircleOptions.prototype.enableClicking;
}
