// angular2-baidu-map/esm2015/types/Control.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function BControlConstructor() { }
function BControlConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (opts: any): BControl
    */
}
/**
 * @record
 */
export function BControl() { }
function BControl_tsickle_Closure_declarations() {
}
/**
 * @record
 */
export function BNavigationControlConstructor() { }
function BNavigationControlConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (opts: BNavigationControlOptions): BNavigationControl
    */
}
/**
 * @record
 */
export function BOverviewMapControlConstructor() { }
function BOverviewMapControlConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (opts: BOverviewMapControlOptions): BOverviewMapControl
    */
}
/**
 * @record
 */
export function BScaleControlConstructor() { }
function BScaleControlConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (opts: BScaleControlOptions): BScaleControl
    */
}
/**
 * @record
 */
export function BMapTypeControlConstructor() { }
function BMapTypeControlConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (opts: BMapTypeControlOptions): BMapTypeControl
    */
}
/**
 * @record
 */
export function BGeolocationConstructor() { }
function BGeolocationConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (opts: BGeolocationControlOptions): BGeolocationControl
    */
}
/**
 * @record
 */
export function BPanoramaControlConstructor() { }
function BPanoramaControlConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (): BPanoramaControlControl
    */
}
/**
 * @record
 */
export function BNavigationControl() { }
function BNavigationControl_tsickle_Closure_declarations() {
    /** @type {?} */
    BNavigationControl.prototype.getType;
    /** @type {?} */
    BNavigationControl.prototype.setType;
}
/**
 * @record
 */
export function BOverviewMapControl() { }
function BOverviewMapControl_tsickle_Closure_declarations() {
    /** @type {?} */
    BOverviewMapControl.prototype.changeView;
    /** @type {?} */
    BOverviewMapControl.prototype.setSize;
    /** @type {?} */
    BOverviewMapControl.prototype.getSize;
}
/**
 * @record
 */
export function BScaleControl() { }
function BScaleControl_tsickle_Closure_declarations() {
    /** @type {?} */
    BScaleControl.prototype.getUnit;
    /** @type {?} */
    BScaleControl.prototype.setUnit;
}
/**
 * @record
 */
export function BMapTypeControl() { }
function BMapTypeControl_tsickle_Closure_declarations() {
}
/**
 * @record
 */
export function BGeolocationControl() { }
function BGeolocationControl_tsickle_Closure_declarations() {
    /** @type {?} */
    BGeolocationControl.prototype.location;
    /** @type {?} */
    BGeolocationControl.prototype.getAddressComponent;
}
/**
 * @record
 */
export function BPanoramaControlControl() { }
function BPanoramaControlControl_tsickle_Closure_declarations() {
}
/**
 * @record
 */
export function ControlOptions() { }
function ControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    ControlOptions.prototype.anchor;
    /** @type {?|undefined} */
    ControlOptions.prototype.offset;
}
/**
 * @record
 */
export function BControlOptions() { }
function BControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    BControlOptions.prototype.anchor;
    /** @type {?|undefined} */
    BControlOptions.prototype.offset;
}
/**
 * @record
 */
export function NavigationControlOptions() { }
function NavigationControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    NavigationControlOptions.prototype.type;
    /** @type {?|undefined} */
    NavigationControlOptions.prototype.showZoomInfo;
    /** @type {?|undefined} */
    NavigationControlOptions.prototype.enableGeolocation;
}
/**
 * @record
 */
export function BNavigationControlOptions() { }
function BNavigationControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    BNavigationControlOptions.prototype.type;
    /** @type {?|undefined} */
    BNavigationControlOptions.prototype.showZoomInfo;
    /** @type {?|undefined} */
    BNavigationControlOptions.prototype.enableGeolocation;
}
/**
 * @record
 */
export function OverviewMapControlOptions() { }
function OverviewMapControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    OverviewMapControlOptions.prototype.size;
    /** @type {?|undefined} */
    OverviewMapControlOptions.prototype.isOpen;
}
/**
 * @record
 */
export function BOverviewMapControlOptions() { }
function BOverviewMapControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    BOverviewMapControlOptions.prototype.size;
    /** @type {?|undefined} */
    BOverviewMapControlOptions.prototype.isOpen;
}
/**
 * @record
 */
export function ScaleControlOptions() { }
function ScaleControlOptions_tsickle_Closure_declarations() {
}
/**
 * @record
 */
export function BScaleControlOptions() { }
function BScaleControlOptions_tsickle_Closure_declarations() {
}
/**
 * @record
 */
export function MapTypeControlOptions() { }
function MapTypeControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    MapTypeControlOptions.prototype.type;
}
/**
 * @record
 */
export function BMapTypeControlOptions() { }
function BMapTypeControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    BMapTypeControlOptions.prototype.type;
}
/**
 * @record
 */
export function GeolocationControlOptions() { }
function GeolocationControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    GeolocationControlOptions.prototype.showAddressBar;
    /** @type {?|undefined} */
    GeolocationControlOptions.prototype.enableAutoLocation;
    /** @type {?|undefined} */
    GeolocationControlOptions.prototype.locationIcon;
}
/**
 * @record
 */
export function BGeolocationControlOptions() { }
function BGeolocationControlOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    BGeolocationControlOptions.prototype.showAddressBar;
    /** @type {?|undefined} */
    BGeolocationControlOptions.prototype.enableAutoLocation;
    /** @type {?|undefined} */
    BGeolocationControlOptions.prototype.locationIcon;
}
/** @enum {number} */
const ControlAnchor = {
    BMAP_ANCHOR_TOP_LEFT: 0,
    BMAP_ANCHOR_TOP_RIGHT: 1,
    BMAP_ANCHOR_BOTTOM_LEFT: 2,
    BMAP_ANCHOR_BOTTOM_RIGHT: 3,
};
export { ControlAnchor };
ControlAnchor[ControlAnchor.BMAP_ANCHOR_TOP_LEFT] = "BMAP_ANCHOR_TOP_LEFT";
ControlAnchor[ControlAnchor.BMAP_ANCHOR_TOP_RIGHT] = "BMAP_ANCHOR_TOP_RIGHT";
ControlAnchor[ControlAnchor.BMAP_ANCHOR_BOTTOM_LEFT] = "BMAP_ANCHOR_BOTTOM_LEFT";
ControlAnchor[ControlAnchor.BMAP_ANCHOR_BOTTOM_RIGHT] = "BMAP_ANCHOR_BOTTOM_RIGHT";
/** @enum {number} */
const NavigationControlType = {
    BMAP_NAVIGATION_CONTROL_LARGE: 0,
    BMAP_NAVIGATION_CONTROL_SMALL: 1,
    BMAP_NAVIGATION_CONTROL_PAN: 2,
    BMAP_NAVIGATION_CONTROL_ZOOM: 3,
};
export { NavigationControlType };
NavigationControlType[NavigationControlType.BMAP_NAVIGATION_CONTROL_LARGE] = "BMAP_NAVIGATION_CONTROL_LARGE";
NavigationControlType[NavigationControlType.BMAP_NAVIGATION_CONTROL_SMALL] = "BMAP_NAVIGATION_CONTROL_SMALL";
NavigationControlType[NavigationControlType.BMAP_NAVIGATION_CONTROL_PAN] = "BMAP_NAVIGATION_CONTROL_PAN";
NavigationControlType[NavigationControlType.BMAP_NAVIGATION_CONTROL_ZOOM] = "BMAP_NAVIGATION_CONTROL_ZOOM";
/** @enum {number} */
const MapTypeControlType = {
    BMAP_MAPTYPE_CONTROL_HORIZONTAL: 0,
    BMAP_MAPTYPE_CONTROL_DROPDOWN: 1,
    BMAP_MAPTYPE_CONTROL_MAP: 2,
};
export { MapTypeControlType };
MapTypeControlType[MapTypeControlType.BMAP_MAPTYPE_CONTROL_HORIZONTAL] = "BMAP_MAPTYPE_CONTROL_HORIZONTAL";
MapTypeControlType[MapTypeControlType.BMAP_MAPTYPE_CONTROL_DROPDOWN] = "BMAP_MAPTYPE_CONTROL_DROPDOWN";
MapTypeControlType[MapTypeControlType.BMAP_MAPTYPE_CONTROL_MAP] = "BMAP_MAPTYPE_CONTROL_MAP";
/** @enum {string} */
const LengthUnit = {
    BMAP_UNIT_METRIC: 'metric',
    BMAP_UNIT_IMPERIAL: 'us',
};
export { LengthUnit };
/**
 * @record
 */
export function AddressComponent() { }
function AddressComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    AddressComponent.prototype.streetNumber;
    /** @type {?} */
    AddressComponent.prototype.street;
    /** @type {?} */
    AddressComponent.prototype.district;
    /** @type {?} */
    AddressComponent.prototype.city;
    /** @type {?} */
    AddressComponent.prototype.province;
}
