// angular2-baidu-map/esm2015/types/Heatmap.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function BHeatmapConstructor() { }
function BHeatmapConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (options?: HeatmapOptions): BHeatmap
    */
}
/**
 * @record
 */
export function BHeatmap() { }
function BHeatmap_tsickle_Closure_declarations() {
    /** @type {?} */
    BHeatmap.prototype.setDataSet;
    /** @type {?} */
    BHeatmap.prototype.setOptions;
}
/**
 * @record
 */
export function HeatmapOptions() { }
function HeatmapOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    HeatmapOptions.prototype.radius;
    /** @type {?|undefined} */
    HeatmapOptions.prototype.visible;
    /** @type {?|undefined} */
    HeatmapOptions.prototype.gradient;
    /** @type {?|undefined} */
    HeatmapOptions.prototype.opacity;
}
/**
 * @record
 */
export function HeatmapPoint() { }
function HeatmapPoint_tsickle_Closure_declarations() {
    /** @type {?} */
    HeatmapPoint.prototype.count;
}
/**
 * @record
 */
export function HeatmapData() { }
function HeatmapData_tsickle_Closure_declarations() {
    /** @type {?} */
    HeatmapData.prototype.data;
    /** @type {?} */
    HeatmapData.prototype.max;
}
