// angular2-baidu-map/esm2015/types/Polyline.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function BPolylineConstructor() { }
function BPolylineConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (points: Array<BPoint>, options?: PolylineOptions): BPolyline
    */
}
/**
 * @record
 */
export function BPolyline() { }
function BPolyline_tsickle_Closure_declarations() {
    /** @type {?} */
    BPolyline.prototype.setPath;
    /** @type {?} */
    BPolyline.prototype.setStrokeColor;
    /** @type {?} */
    BPolyline.prototype.setStrokeOpacity;
    /** @type {?} */
    BPolyline.prototype.setStrokeWeight;
    /** @type {?} */
    BPolyline.prototype.setStrokeStyle;
    /** @type {?} */
    BPolyline.prototype.enableEditing;
    /** @type {?} */
    BPolyline.prototype.disableEditing;
    /** @type {?} */
    BPolyline.prototype.enableMassClear;
    /** @type {?} */
    BPolyline.prototype.disableMassClear;
}
/**
 * @record
 */
export function PolylineOptions() { }
function PolylineOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    PolylineOptions.prototype.strokeColor;
    /** @type {?|undefined} */
    PolylineOptions.prototype.strokeWeight;
    /** @type {?|undefined} */
    PolylineOptions.prototype.strokeOpacity;
    /** @type {?|undefined} */
    PolylineOptions.prototype.strokeStyle;
    /** @type {?|undefined} */
    PolylineOptions.prototype.enableMassClear;
    /** @type {?|undefined} */
    PolylineOptions.prototype.enableEditing;
    /** @type {?|undefined} */
    PolylineOptions.prototype.enableClicking;
    /** @type {?|undefined} */
    PolylineOptions.prototype.icons;
}
