// angular2-baidu-map/esm2015/types/Polygon.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function BPolygonConstructor() { }
function BPolygonConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (points: Array<BPoint>, options?: PolygonOptions): BPolygon
    */
}
/**
 * @record
 */
export function BPolygon() { }
function BPolygon_tsickle_Closure_declarations() {
    /** @type {?} */
    BPolygon.prototype.setPath;
    /** @type {?} */
    BPolygon.prototype.setStrokeColor;
    /** @type {?} */
    BPolygon.prototype.setFillColor;
    /** @type {?} */
    BPolygon.prototype.setStrokeOpacity;
    /** @type {?} */
    BPolygon.prototype.setStrokeWeight;
    /** @type {?} */
    BPolygon.prototype.setFillOpacity;
    /** @type {?} */
    BPolygon.prototype.setStrokeStyle;
    /** @type {?} */
    BPolygon.prototype.enableEditing;
    /** @type {?} */
    BPolygon.prototype.disableEditing;
    /** @type {?} */
    BPolygon.prototype.enableMassClear;
    /** @type {?} */
    BPolygon.prototype.disableMassClear;
}
/**
 * @record
 */
export function PolygonOptions() { }
function PolygonOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    PolygonOptions.prototype.strokeColor;
    /** @type {?|undefined} */
    PolygonOptions.prototype.fillColor;
    /** @type {?|undefined} */
    PolygonOptions.prototype.strokeWeight;
    /** @type {?|undefined} */
    PolygonOptions.prototype.strokeOpacity;
    /** @type {?|undefined} */
    PolygonOptions.prototype.fillOpacity;
    /** @type {?|undefined} */
    PolygonOptions.prototype.strokeStyle;
    /** @type {?|undefined} */
    PolygonOptions.prototype.enableMassClear;
    /** @type {?|undefined} */
    PolygonOptions.prototype.enableEditing;
    /** @type {?|undefined} */
    PolygonOptions.prototype.enableClicking;
}
