// angular2-baidu-map/esm2015/types/TileLayer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function BTileLayerConstructor() { }
function BTileLayerConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (options?: TileLayerOptions): BTileLayer
    */
}
/**
 * @record
 */
export function BTileLayer() { }
function BTileLayer_tsickle_Closure_declarations() {
    /** @type {?} */
    BTileLayer.prototype.getTilesUrl;
    /** @type {?} */
    BTileLayer.prototype.getCopyright;
    /** @type {?} */
    BTileLayer.prototype.isTransparentPng;
}
/**
 * @record
 */
export function TileLayerOptions() { }
function TileLayerOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    TileLayerOptions.prototype.transparentPng;
    /** @type {?|undefined} */
    TileLayerOptions.prototype.tileUrlTemplate;
    /** @type {?|undefined} */
    TileLayerOptions.prototype.copyright;
    /** @type {?|undefined} */
    TileLayerOptions.prototype.zIndex;
}
/**
 * @record
 */
export function getTilesUrlFunc() { }
function getTilesUrlFunc_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    (tileCoord: BPixel, zoom: number): string
    */
}
