// angular2-baidu-map/esm5/components/heatmap.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { nullCheck } from '../helpers/validate';
import { MapService } from '../providers/mapService';
import { ScriptLoader } from '../providers/scriptLoader';
var /** @type {?} */ LIB_URL = "https://api.map.baidu.com/library/Heatmap/2.0/src/Heatmap_min.js";
var HeatmapComponent = /** @class */ (function () {
    function HeatmapComponent(_service, scriptLoader) {
        this._service = _service;
        this.scriptLoader = scriptLoader;
        this.loaded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    HeatmapComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        nullCheck(this.dataset, 'dataset is required for <heatmap>');
        this._service.getNativeMap().then(function () {
            return _this.scriptLoader.load(LIB_URL, false, function () {
                _this._service
                    .addOverlay(function () {
                    return (_this.heatmap = new window.BMapLib.HeatmapOverlay(_this.options));
                })
                    .then(function () {
                    _this.loaded.emit(_this.heatmap);
                    if (_this.dataset) {
                        _this.heatmap.setDataSet(_this.dataset);
                    }
                });
            });
        });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    HeatmapComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes["dataset"] && !changes["dataset"].isFirstChange()) {
            this.heatmap.setDataSet(changes["dataset"].currentValue);
        }
        if (changes["options"] && !changes["options"].isFirstChange()) {
            this.heatmap.setOptions(changes["options"].currentValue);
        }
    };
    /**
     * @return {?}
     */
    HeatmapComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._service.removeOverlay(this.heatmap);
    };
    HeatmapComponent.decorators = [
        { type: Directive, args: [{
                    selector: 'heatmap'
                },] },
    ];
    /** @nocollapse */
    HeatmapComponent.ctorParameters = function () { return [
        { type: MapService, },
        { type: ScriptLoader, },
    ]; };
    HeatmapComponent.propDecorators = {
        "dataset": [{ type: Input },],
        "options": [{ type: Input },],
        "loaded": [{ type: Output },],
    };
    return HeatmapComponent;
}());
export { HeatmapComponent };
function HeatmapComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    HeatmapComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    HeatmapComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    HeatmapComponent.propDecorators;
    /** @type {?} */
    HeatmapComponent.prototype.dataset;
    /** @type {?} */
    HeatmapComponent.prototype.options;
    /** @type {?} */
    HeatmapComponent.prototype.loaded;
    /** @type {?} */
    HeatmapComponent.prototype.heatmap;
    /** @type {?} */
    HeatmapComponent.prototype._service;
    /** @type {?} */
    HeatmapComponent.prototype.scriptLoader;
}
