// angular2-baidu-map/esm5/components/tilelayer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { MapService } from '../providers/mapService';
var TileLayderComponent = /** @class */ (function () {
    function TileLayderComponent(_service) {
        this._service = _service;
        this.loaded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TileLayderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ func = this.getTilesUrl;
        this._service
            .addTileLayer(function () {
            _this.tilelayer = new window.BMap.TileLayer(_this.options);
            if (_this.getTilesUrl) {
                _this.tilelayer.getTilesUrl = function (tileCoord, zoom) {
                    return func(tileCoord, zoom);
                };
            }
            return _this.tilelayer;
        })
            .then(function () {
            _this.loaded.emit(_this.tilelayer);
        });
    };
    /**
     * @return {?}
     */
    TileLayderComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._service.removeTileLayer(this.tilelayer);
    };
    TileLayderComponent.decorators = [
        { type: Directive, args: [{
                    selector: 'tilelayer'
                },] },
    ];
    /** @nocollapse */
    TileLayderComponent.ctorParameters = function () { return [
        { type: MapService, },
    ]; };
    TileLayderComponent.propDecorators = {
        "getTilesUrl": [{ type: Input },],
        "options": [{ type: Input },],
        "loaded": [{ type: Output },],
    };
    return TileLayderComponent;
}());
export { TileLayderComponent };
function TileLayderComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TileLayderComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TileLayderComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    TileLayderComponent.propDecorators;
    /** @type {?} */
    TileLayderComponent.prototype.getTilesUrl;
    /** @type {?} */
    TileLayderComponent.prototype.options;
    /** @type {?} */
    TileLayderComponent.prototype.loaded;
    /** @type {?} */
    TileLayderComponent.prototype.tilelayer;
    /** @type {?} */
    TileLayderComponent.prototype._service;
}
