// angular2-baidu-map/esm5/components/trafficlayer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { MapService } from '../providers/mapService';
var TrafficLayderComponent = /** @class */ (function () {
    function TrafficLayderComponent(_service) {
        this._service = _service;
        this.loaded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TrafficLayderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._service
            .addTileLayer(function () {
            _this.trafficlayer = new window.BMap.TrafficLayer(_this.options);
            return _this.trafficlayer;
        })
            .then(function () {
            _this.loaded.emit(_this.trafficlayer);
        });
    };
    /**
     * @return {?}
     */
    TrafficLayderComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._service.removeTileLayer(this.trafficlayer);
    };
    TrafficLayderComponent.decorators = [
        { type: Directive, args: [{
                    selector: 'trafficlayer'
                },] },
    ];
    /** @nocollapse */
    TrafficLayderComponent.ctorParameters = function () { return [
        { type: MapService, },
    ]; };
    TrafficLayderComponent.propDecorators = {
        "options": [{ type: Input },],
        "loaded": [{ type: Output },],
    };
    return TrafficLayderComponent;
}());
export { TrafficLayderComponent };
function TrafficLayderComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TrafficLayderComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TrafficLayderComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    TrafficLayderComponent.propDecorators;
    /** @type {?} */
    TrafficLayderComponent.prototype.options;
    /** @type {?} */
    TrafficLayderComponent.prototype.loaded;
    /** @type {?} */
    TrafficLayderComponent.prototype.trafficlayer;
    /** @type {?} */
    TrafficLayderComponent.prototype._service;
}
