// angular2-baidu-map/esm5/types/MapType.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function BMapTypeConstructor() { }
function BMapTypeConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (name: string, layers: BTileLayer | Array<BTileLayer>, options: BMapTypeOptions): BMapType
    */
}
/**
 * @record
 */
export function BMapTypeOptions() { }
function BMapTypeOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    BMapTypeOptions.prototype.minZoom;
    /** @type {?} */
    BMapTypeOptions.prototype.maxZoom;
    /** @type {?} */
    BMapTypeOptions.prototype.errorImageUrl;
    /** @type {?} */
    BMapTypeOptions.prototype.textColor;
    /** @type {?} */
    BMapTypeOptions.prototype.tips;
}
/**
 * @record
 */
export function BMapType() { }
function BMapType_tsickle_Closure_declarations() {
    /** @type {?} */
    BMapType.prototype.getName;
    /** @type {?} */
    BMapType.prototype.getTileLayer;
    /** @type {?} */
    BMapType.prototype.getMinZoom;
    /** @type {?} */
    BMapType.prototype.getMaxZoom;
    /** @type {?} */
    BMapType.prototype.getProjection;
    /** @type {?} */
    BMapType.prototype.getTextColor;
    /** @type {?} */
    BMapType.prototype.getTips;
}
