// angular2-baidu-map/esm5/types/MarkerClusterer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function BMarkerClustererConstructor() { }
function BMarkerClustererConstructor_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    new (map: BMapInstance, options?: BMarkerClustererOptions): BMarkerClusterer
    */
}
/**
 * @record
 */
export function BMarkerClusterer() { }
function BMarkerClusterer_tsickle_Closure_declarations() {
    /** @type {?} */
    BMarkerClusterer.prototype.addMarker;
    /** @type {?} */
    BMarkerClusterer.prototype.addMarkers;
    /** @type {?} */
    BMarkerClusterer.prototype.clearMarkers;
    /** @type {?} */
    BMarkerClusterer.prototype.getClustersCount;
    /** @type {?} */
    BMarkerClusterer.prototype.getGridSize;
    /** @type {?} */
    BMarkerClusterer.prototype.getMap;
    /** @type {?} */
    BMarkerClusterer.prototype.getMarkers;
    /** @type {?} */
    BMarkerClusterer.prototype.getMaxZoom;
    /** @type {?} */
    BMarkerClusterer.prototype.getMinClusterSize;
    /** @type {?} */
    BMarkerClusterer.prototype.getStyles;
    /** @type {?} */
    BMarkerClusterer.prototype.isAverageCenter;
    /** @type {?} */
    BMarkerClusterer.prototype.removeMarker;
    /** @type {?} */
    BMarkerClusterer.prototype.removeMarkers;
    /** @type {?} */
    BMarkerClusterer.prototype.setGridSize;
    /** @type {?} */
    BMarkerClusterer.prototype.setMaxZoom;
    /** @type {?} */
    BMarkerClusterer.prototype.setMinClusterSize;
    /** @type {?} */
    BMarkerClusterer.prototype.setStyles;
}
/**
 * @record
 */
export function BMarkerClustererOptions() { }
function BMarkerClustererOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    BMarkerClustererOptions.prototype.markers;
    /** @type {?|undefined} */
    BMarkerClustererOptions.prototype.girdSize;
    /** @type {?|undefined} */
    BMarkerClustererOptions.prototype.maxZoom;
    /** @type {?|undefined} */
    BMarkerClustererOptions.prototype.minClusterSize;
    /** @type {?|undefined} */
    BMarkerClustererOptions.prototype.isAverangeCenter;
    /** @type {?|undefined} */
    BMarkerClustererOptions.prototype.styles;
}
/**
 * @record
 */
export function MarkerClustererOptions() { }
function MarkerClustererOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    MarkerClustererOptions.prototype.markers;
    /** @type {?|undefined} */
    MarkerClustererOptions.prototype.girdSize;
    /** @type {?|undefined} */
    MarkerClustererOptions.prototype.maxZoom;
    /** @type {?|undefined} */
    MarkerClustererOptions.prototype.minClusterSize;
    /** @type {?|undefined} */
    MarkerClustererOptions.prototype.isAverangeCenter;
    /** @type {?|undefined} */
    MarkerClustererOptions.prototype.styles;
}
