// angular2-baidu-map/esm2015/components/control.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { toGeolocationOptions, toMapTypeControlOptions, toNavigationControlOptions, toOverviewMapControlOptions, toScaleControlOptions } from '../helpers/transformer';
import { nullCheck } from '../helpers/validate';
import { MapService } from '../providers/mapService';
export class ControlComponent {
    /**
     * @param {?} _service
     */
    constructor(_service) {
        this._service = _service;
        this.loaded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        nullCheck(this.type, 'type is required for <control>');
        this._service
            .addControl(() => {
            this.control = this.createControl(this.type, this.options);
            return this.control;
        })
            .then(({ control }) => {
            this.loaded.emit(control);
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._service.removeControl(this.control);
    }
    /**
     * @param {?} type
     * @param {?} options
     * @return {?}
     */
    createControl(type, options) {
        if (type === 'navigation') {
            return new window.BMap.NavigationControl(toNavigationControlOptions(options));
        }
        if (type === 'overviewmap') {
            return new window.BMap.OverviewMapControl(toOverviewMapControlOptions(options));
        }
        if (type === 'scale') {
            return new window.BMap.ScaleControl(toScaleControlOptions(options));
        }
        if (type === 'maptype') {
            return new window.BMap.MapTypeControl(toMapTypeControlOptions(options));
        }
        if (type === 'geolocation') {
            return new window.BMap.GeolocationControl(toGeolocationOptions(options));
        }
        if (type === 'panorama') {
            return new window.BMap.PanoramaControl();
        }
        throw new Error(`Unsupported type:${type} of control`);
    }
}
ControlComponent.decorators = [
    { type: Directive, args: [{
                selector: 'control'
            },] },
];
/** @nocollapse */
ControlComponent.ctorParameters = () => [
    { type: MapService, },
];
ControlComponent.propDecorators = {
    "type": [{ type: Input },],
    "options": [{ type: Input },],
    "loaded": [{ type: Output },],
};
function ControlComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ControlComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ControlComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    ControlComponent.propDecorators;
    /** @type {?} */
    ControlComponent.prototype.type;
    /** @type {?} */
    ControlComponent.prototype.options;
    /** @type {?} */
    ControlComponent.prototype.loaded;
    /** @type {?} */
    ControlComponent.prototype.control;
    /** @type {?} */
    ControlComponent.prototype._service;
}
