// angular2-baidu-map/esm2015/components/heatmap.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { nullCheck } from '../helpers/validate';
import { MapService } from '../providers/mapService';
import { ScriptLoader } from '../providers/scriptLoader';
const /** @type {?} */ LIB_URL = `https://api.map.baidu.com/library/Heatmap/2.0/src/Heatmap_min.js`;
export class HeatmapComponent {
    /**
     * @param {?} _service
     * @param {?} scriptLoader
     */
    constructor(_service, scriptLoader) {
        this._service = _service;
        this.scriptLoader = scriptLoader;
        this.loaded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        nullCheck(this.dataset, 'dataset is required for <heatmap>');
        this._service.getNativeMap().then(() => {
            return this.scriptLoader.load(LIB_URL, false, () => {
                this._service
                    .addOverlay(() => {
                    return (this.heatmap = new window.BMapLib.HeatmapOverlay(this.options));
                })
                    .then(() => {
                    this.loaded.emit(this.heatmap);
                    if (this.dataset) {
                        this.heatmap.setDataSet(this.dataset);
                    }
                });
            });
        });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes["dataset"] && !changes["dataset"].isFirstChange()) {
            this.heatmap.setDataSet(changes["dataset"].currentValue);
        }
        if (changes["options"] && !changes["options"].isFirstChange()) {
            this.heatmap.setOptions(changes["options"].currentValue);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._service.removeOverlay(this.heatmap);
    }
}
HeatmapComponent.decorators = [
    { type: Directive, args: [{
                selector: 'heatmap'
            },] },
];
/** @nocollapse */
HeatmapComponent.ctorParameters = () => [
    { type: MapService, },
    { type: ScriptLoader, },
];
HeatmapComponent.propDecorators = {
    "dataset": [{ type: Input },],
    "options": [{ type: Input },],
    "loaded": [{ type: Output },],
};
function HeatmapComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    HeatmapComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    HeatmapComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    HeatmapComponent.propDecorators;
    /** @type {?} */
    HeatmapComponent.prototype.dataset;
    /** @type {?} */
    HeatmapComponent.prototype.options;
    /** @type {?} */
    HeatmapComponent.prototype.loaded;
    /** @type {?} */
    HeatmapComponent.prototype.heatmap;
    /** @type {?} */
    HeatmapComponent.prototype._service;
    /** @type {?} */
    HeatmapComponent.prototype.scriptLoader;
}
