// angular2-baidu-map/esm2015/components/map.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { MapService } from '../providers/mapService';
import { nullCheck } from '../helpers/validate';
export class MapComponent {
    /**
     * @param {?} _service
     */
    constructor(_service) {
        this._service = _service;
        this.loaded = new EventEmitter();
        this.clicked = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        nullCheck(this.options, 'options is required for <baidu-map>');
        nullCheck(this.options.centerAndZoom, 'options.centerAndZoom is required for <baidu-map>');
        this._service
            .createMap(this.mapInstance.nativeElement, this.options)
            .then(map => {
            this.loaded.emit(map);
            return map;
        })
            .then(map => {
            this.addListener(map);
        });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        const /** @type {?} */ opts = /** @type {?} */ (changes["options"].currentValue);
        if (!opts) {
            return console.warn('MapOptions change was ignored since you are passing empty value');
        }
        this._service.setOptions(opts);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        console.log('on map destroy');
    }
    /**
     * @param {?} map
     * @return {?}
     */
    addListener(map) {
        map.addEventListener('click', (e) => {
            this.clicked.emit(e);
        });
    }
}
MapComponent.decorators = [
    { type: Component, args: [{
                providers: [MapService],
                selector: 'baidu-map',
                styles: [
                    `
        .baidu-map-instance {
            width: 100%;
            height: 100%;
            display: none;
        }
        .baidu-map-offline {
            width: 100%;
            height: 100%;
            background-color: #E6E6E6;
            position: relative;
            display: none;
        }
        .baidu-map-offline label {
            fontSize: 30px;
            margin: 0;
            position: absolute;
            top: 50%;
            left: 50%;
            margin-right: -50%;
            transform: translate(-50%, -50%);
        }
        .tranclude-content {
            display: none;
        }
        `
                ],
                template: `
    <div #mapinstance class="baidu-map-instance"></div>
    <div class="baidu-map-offline">
        <label>{{ 'NO_NETWORK' }}</label>
    </div>
    <div class="tranclude-content">
        <ng-content></ng-content>
    </div>
    `
            },] },
];
/** @nocollapse */
MapComponent.ctorParameters = () => [
    { type: MapService, },
];
MapComponent.propDecorators = {
    "options": [{ type: Input },],
    "loaded": [{ type: Output },],
    "clicked": [{ type: Output },],
    "mapInstance": [{ type: ViewChild, args: ['mapinstance',] },],
};
function MapComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MapComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MapComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    MapComponent.propDecorators;
    /** @type {?} */
    MapComponent.prototype.options;
    /** @type {?} */
    MapComponent.prototype.loaded;
    /** @type {?} */
    MapComponent.prototype.clicked;
    /** @type {?} */
    MapComponent.prototype.mapInstance;
    /** @type {?} */
    MapComponent.prototype._service;
}
