// angular2-baidu-map/esm2015/components/tilelayer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { MapService } from '../providers/mapService';
export class TileLayderComponent {
    /**
     * @param {?} _service
     */
    constructor(_service) {
        this._service = _service;
        this.loaded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        const /** @type {?} */ func = this.getTilesUrl;
        this._service
            .addTileLayer(() => {
            this.tilelayer = new window.BMap.TileLayer(this.options);
            if (this.getTilesUrl) {
                this.tilelayer.getTilesUrl = function (tileCoord, zoom) {
                    return func(tileCoord, zoom);
                };
            }
            return this.tilelayer;
        })
            .then(() => {
            this.loaded.emit(this.tilelayer);
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._service.removeTileLayer(this.tilelayer);
    }
}
TileLayderComponent.decorators = [
    { type: Directive, args: [{
                selector: 'tilelayer'
            },] },
];
/** @nocollapse */
TileLayderComponent.ctorParameters = () => [
    { type: MapService, },
];
TileLayderComponent.propDecorators = {
    "getTilesUrl": [{ type: Input },],
    "options": [{ type: Input },],
    "loaded": [{ type: Output },],
};
function TileLayderComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TileLayderComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TileLayderComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    TileLayderComponent.propDecorators;
    /** @type {?} */
    TileLayderComponent.prototype.getTilesUrl;
    /** @type {?} */
    TileLayderComponent.prototype.options;
    /** @type {?} */
    TileLayderComponent.prototype.loaded;
    /** @type {?} */
    TileLayderComponent.prototype.tilelayer;
    /** @type {?} */
    TileLayderComponent.prototype._service;
}
