// angular2-baidu-map/esm2015/components/trafficlayer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { MapService } from '../providers/mapService';
export class TrafficLayderComponent {
    /**
     * @param {?} _service
     */
    constructor(_service) {
        this._service = _service;
        this.loaded = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._service
            .addTileLayer(() => {
            this.trafficlayer = new window.BMap.TrafficLayer(this.options);
            return this.trafficlayer;
        })
            .then(() => {
            this.loaded.emit(this.trafficlayer);
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._service.removeTileLayer(this.trafficlayer);
    }
}
TrafficLayderComponent.decorators = [
    { type: Directive, args: [{
                selector: 'trafficlayer'
            },] },
];
/** @nocollapse */
TrafficLayderComponent.ctorParameters = () => [
    { type: MapService, },
];
TrafficLayderComponent.propDecorators = {
    "options": [{ type: Input },],
    "loaded": [{ type: Output },],
};
function TrafficLayderComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    TrafficLayderComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    TrafficLayderComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    TrafficLayderComponent.propDecorators;
    /** @type {?} */
    TrafficLayderComponent.prototype.options;
    /** @type {?} */
    TrafficLayderComponent.prototype.loaded;
    /** @type {?} */
    TrafficLayderComponent.prototype.trafficlayer;
    /** @type {?} */
    TrafficLayderComponent.prototype._service;
}
