// angular-gridster2/esm2015/lib/gridsterConfigS.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function GridsterConfigS() { }
function GridsterConfigS_tsickle_Closure_declarations() {
    /** @type {?} */
    GridsterConfigS.prototype.gridType;
    /** @type {?} */
    GridsterConfigS.prototype.fixedColWidth;
    /** @type {?} */
    GridsterConfigS.prototype.fixedRowHeight;
    /** @type {?} */
    GridsterConfigS.prototype.keepFixedHeightInMobile;
    /** @type {?} */
    GridsterConfigS.prototype.keepFixedWidthInMobile;
    /** @type {?} */
    GridsterConfigS.prototype.setGridSize;
    /** @type {?} */
    GridsterConfigS.prototype.compactType;
    /** @type {?} */
    GridsterConfigS.prototype.mobileBreakpoint;
    /** @type {?} */
    GridsterConfigS.prototype.minCols;
    /** @type {?} */
    GridsterConfigS.prototype.maxCols;
    /** @type {?} */
    GridsterConfigS.prototype.minRows;
    /** @type {?} */
    GridsterConfigS.prototype.maxRows;
    /** @type {?} */
    GridsterConfigS.prototype.defaultItemCols;
    /** @type {?} */
    GridsterConfigS.prototype.defaultItemRows;
    /** @type {?} */
    GridsterConfigS.prototype.maxItemCols;
    /** @type {?} */
    GridsterConfigS.prototype.maxItemRows;
    /** @type {?} */
    GridsterConfigS.prototype.minItemCols;
    /** @type {?} */
    GridsterConfigS.prototype.minItemRows;
    /** @type {?} */
    GridsterConfigS.prototype.minItemArea;
    /** @type {?} */
    GridsterConfigS.prototype.maxItemArea;
    /** @type {?} */
    GridsterConfigS.prototype.margin;
    /** @type {?} */
    GridsterConfigS.prototype.outerMargin;
    /** @type {?} */
    GridsterConfigS.prototype.outerMarginTop;
    /** @type {?} */
    GridsterConfigS.prototype.outerMarginRight;
    /** @type {?} */
    GridsterConfigS.prototype.outerMarginBottom;
    /** @type {?} */
    GridsterConfigS.prototype.outerMarginLeft;
    /** @type {?} */
    GridsterConfigS.prototype.scrollSensitivity;
    /** @type {?} */
    GridsterConfigS.prototype.scrollSpeed;
    /** @type {?} */
    GridsterConfigS.prototype.draggable;
    /** @type {?} */
    GridsterConfigS.prototype.resizable;
    /** @type {?} */
    GridsterConfigS.prototype.swap;
    /** @type {?} */
    GridsterConfigS.prototype.pushItems;
    /** @type {?} */
    GridsterConfigS.prototype.disablePushOnDrag;
    /** @type {?} */
    GridsterConfigS.prototype.disablePushOnResize;
    /** @type {?} */
    GridsterConfigS.prototype.disableAutoPositionOnConflict;
    /** @type {?} */
    GridsterConfigS.prototype.pushDirections;
    /** @type {?} */
    GridsterConfigS.prototype.pushResizeItems;
    /** @type {?} */
    GridsterConfigS.prototype.displayGrid;
    /** @type {?} */
    GridsterConfigS.prototype.disableWindowResize;
    /** @type {?} */
    GridsterConfigS.prototype.disableWarnings;
    /** @type {?} */
    GridsterConfigS.prototype.scrollToNewItems;
    /** @type {?} */
    GridsterConfigS.prototype.enableEmptyCellClick;
    /** @type {?} */
    GridsterConfigS.prototype.enableEmptyCellContextMenu;
    /** @type {?} */
    GridsterConfigS.prototype.enableEmptyCellDrop;
    /** @type {?} */
    GridsterConfigS.prototype.enableEmptyCellDrag;
    /** @type {?} */
    GridsterConfigS.prototype.emptyCellDragMaxCols;
    /** @type {?} */
    GridsterConfigS.prototype.emptyCellDragMaxRows;
    /** @type {?} */
    GridsterConfigS.prototype.ignoreMarginInRow;
    /** @type {?} */
    GridsterConfigS.prototype.api;
    /* TODO: handle strange member:
    [propName: string]: any;
    */
}
/**
 * @record
 */
export function DragBase() { }
function DragBase_tsickle_Closure_declarations() {
    /** @type {?} */
    DragBase.prototype.enabled;
    /** @type {?} */
    DragBase.prototype.delayStart;
    /* TODO: handle strange member:
    [propName: string]: any;
    */
}
/**
 * @record
 */
export function Draggable() { }
function Draggable_tsickle_Closure_declarations() {
    /** @type {?} */
    Draggable.prototype.ignoreContentClass;
    /** @type {?} */
    Draggable.prototype.ignoreContent;
    /** @type {?} */
    Draggable.prototype.dragHandleClass;
    /** @type {?} */
    Draggable.prototype.dropOverItems;
    /** @type {?} */
    Draggable.prototype.dropOverItemsCallback;
}
/**
 * @record
 */
export function Resizable() { }
function Resizable_tsickle_Closure_declarations() {
    /** @type {?} */
    Resizable.prototype.handles;
}
/**
 * @record
 */
export function PushDirections() { }
function PushDirections_tsickle_Closure_declarations() {
    /** @type {?} */
    PushDirections.prototype.north;
    /** @type {?} */
    PushDirections.prototype.east;
    /** @type {?} */
    PushDirections.prototype.south;
    /** @type {?} */
    PushDirections.prototype.west;
}
